Extend documentation for deploy development (#260)
civiform/docs: docs/contributor-guide/developer-guide/deploy-system/developer-guide.md

# Deploy system developer guide

This page contains information for deploy system developers, which go beyond the initial setup. 
You should have worked through the [Prerequisits & dev setup](contributor-guide/developer-guide/deploy-system/prerequisits.md) already

> and know the difference between the repositories (civiform-deploy, civiform-deploy-infra and civifor) from the overview section.

Reviewer: ```suggestion
and know the difference between the repositories (civiform-deploy, civiform-deploy-infra and civiform) from the overview section.
```

Author: Done

The "Most common developer flows" section contains all essential information to be able to do a first change. 
We recommend coming back soon to "Optimizing your flow" and skim reading the tips and tricks section so you know 
what information is provided when you need it.

## Running locally

The most common flow for developers is to run the deployment script on your local machine to test your changes to 

> the civiform-deploy-infra repository. To do this,you create a configuration file for your deployment and then run 

Reviewer 2: I wonder if this should actually be in prerequisites. Or maybe there's another way to more clearly delineate between the one time setup work we need to do (setting up AWS account, creating config file, etc.), and the steps for ongoing development.

Reviewer 2: gotcha, thank you for explaining!

the setup script which pulls in the configuration values and applies them via terraform to the deployment on your 
chosen cloud provider (AWS/Azure).

You have two options here: 

1. The first flow is similar to what a civic entity would do. From within the civiform-deploy repository you run
[bin/setup](https://github.com/civiform/civiform-deploy/blob/main/bin/setup), which then calls 
run.py(https://github.com/civiform/cloud-deploy-infra/blob/main/cloud/shared/bin/run.py). 
Because run.py is in a different repository, it requires development across different git repositories.

> 2. For the second flow you develope in the civiform-deploy-infra repository and run run.py from there.

Reviewer 2: ```suggestion
2. For the second flow, you develop in the civiform-deploy-infra repository and run run.py from there.
```

You can use the [AWS](https://github.com/civiform/civiform-staging-deploy/blob/main/aws_staging_civiform_config.sh) and 
[Azure](https://github.com/civiform/civiform-staging-deploy/blob/main/azure_staging_civiform_config.sh)
config files as additional examples.

TODO(#4324) complete the list of values that need to be changed as a minimal set for deployment to work (step 2).

### Flow 1: Running locally from within civiform-deploy (like operations at a civic entity would)

Once you have adapted the configuration file you can run the deployment script. 
Run the script from your civiform-deploy repository's root directory:
```
[bin/setup](https://github.com/civiform/civiform-deploy/blob/main/bin/setup)
```
Setup will ask you to confirm multiple times and to provide secret values. You can enter random values and click yes as needed.
At the end the script should print url to your dev server!

Remember that if you have any changes in your civiform-deploy-infra repository which you would like to be visible in your deployment, 
you have to have gone through the git setup steps in the prerequisits previously to ensure that the newest version is copied

> from your civiform-deploy-infra to your civiform-deploy repository. Also make sure commit the changes to the relevant 

Reviewer 2: ```suggestion
from your civiform-deploy-infra to your civiform-deploy repository. Also make sure to commit the changes to the relevant 
```

branch in git before running the script.

> ### Flow 2 Running the deploy script from within civiform-deploy-infra

Reviewer 2: Very cool! I haven't tried this option yet.

If you don't have a dockerhub acount yet, you can register [here] {https://hub.docker.com/}.
* Then see your dockerhub at https://hub.docker.com/u/<your_dockerhub_name>, 
* Click on Repositories at the top
* Click on "Create Repository"
* Follow the steps, in the following we'll presume that you named your repository "civiform" so that it will be available at <your_dockerhub_name>/civiform

#### Building your docker image locally

In your civiform repository, 
* open bin/build-prod
* change "civiform/" to "<your_dockerhub_name> wherever you find it
* run bin/build-prod

Push the image to your repository
* check the image is present with '''docker images'''
* push it to docker with '''docker push <your_cockerhub_name>/civiform:latest'''. (If you get an access denied error, you may need to log into your docker account with '''docker login''')
* go to your dockerhub at https://hub.docker.com/u/<your_dockerhub_name> to check the image has been updated

#### Pointing your deployment at your docker image
* Open app.tf (https://github.com/civiform/cloud-deploy-infra/blob/main/cloud/aws/templates/aws_oidc/app.tf#L37) 

> * Change the container image to '''container_image              = “jessbumblebee/civiform:latest” '''

Reviewer 2: ```suggestion
* Change the container image to '''container_image              = “<your_github_username>/civiform:latest” '''
```
